### PR TITLE
GEODE-8874: Fix for transaction commit failure

### PR DESCRIPTION
**Pull request by jvarenina**
Issue: Transaction commit fail when m_transactionID integer overflows to
negative value, because server interprets negative value as
non-transactional message.

Fix: Client is impacted to reset transactionID to one when INT_32MAX
value is reached in order to avoid overflow.

### CODE REVIEW

File: cppcache/src/TXId.cpp [hunk 1]
```diff
@@ -27,15 +27,27 @@ namespace apache {
...
-std::atomic<int32_t> TXId::m_transactionId(1);
+std::atomic<int32_t> TXId::m_transactionId(0);
 
-TXId::TXId() : m_TXId(m_transactionId++) {}
+TXId::TXId() {
+  // If m_transactionId has reached maximum value, then set
+  // it to zero in order to avoid overflow to negative value.
+  auto maxValueTXId = INT32_MAX;
+  m_transactionId.compare_exchange_strong(maxValueTXId, 0);
+  m_TXId = ++m_transactionId;
```
Comment (pivotal-jbarrett) [inline cppcache/src/TXId.cpp:37]: While highly unlikely it is possible for `m_transactionId` to be at max value again between lines 36 and 37. This whole operation should be done in a loop where you fetch the current, increment if less than max or reset to 0, then compare and set the new value. If failure to set, then try again.

```c++
auto current = m_transactionId.load();
decltype(current) next;
do {
    next = std::numeric_limits<decltype(current)>::max() == current ? 1 : current + 1;
} while(!m_transactionId.compare_exchange_strong(current, next)); 

m_TXId = next;
```

File: cppcache/src/TXId.cpp [hunk 2]
```diff
@@ -27,15 +27,27 @@ namespace apache {
...
 namespace client {
 
-std::atomic<int32_t> TXId::m_transactionId(1);
+std::atomic<int32_t> TXId::m_transactionId(0);
 
-TXId::TXId() : m_TXId(m_transactionId++) {}
+TXId::TXId() {
+  // If m_transactionId has reached maximum value, then set
+  // it to zero in order to avoid overflow to negative value.
+  auto maxValueTXId = INT32_MAX;
```
Comment (pivotal-jbarrett) [inline cppcache/src/TXId.cpp:35]:
```c++
auto maxValueTXId = std::numeric_limits<int>::max();
```
https://en.cppreference.com/w/cpp/types/numeric_limits

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

File: cppcache/src/TXId.cpp (hunk 1 above)
Comment (jvarenina) [reply cppcache/src/TXId.cpp]: You are right. I have updated solution as you suggested.

File: cppcache/src/TXId.cpp (hunk 2 above)
Comment (jvarenina) [reply cppcache/src/TXId.cpp]: Thanks, done.

File: cppcache/src/TXId.cpp [hunk 3]
```diff
@@ -27,15 +27,34 @@ namespace apache {
...
 
-std::atomic<int32_t> TXId::m_transactionId(1);
+std::atomic<int32_t> TXId::m_transactionId(0);
 
-TXId::TXId() : m_TXId(m_transactionId++) {}
+TXId::TXId() {
+  // If m_transactionId has reached maximum value, then start again
+  // from zero, and increment by one until first unused value is found.
+  // This is done to avoid overflow of m_transactionId to negative value.
+  auto maxValueTXId = std::numeric_limits<int>::max();
```
Comment (pivotal-jbarrett) [inline cppcache/src/TXId.cpp:36]: Unused.

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

**Review by pivotal-jbarrett — APPROVED** [review]

File: cppcache/test/TXIdTest.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+#include <gtest/gtest.h>
+
+namespace apache {
+namespace geode {
+namespace client {
+
+TEST(TXIdTest, testIncrementTransactionID) {
+  auto tx = new TXId();
+  ASSERT_EQ(1, tx->getId());
+
```
Comment (pdxcodemonkey) [inline cppcache/test/TXIdTest.cpp:29]: need `delete(tx);` ?

**Review by pdxcodemonkey — CHANGES_REQUESTED** [review]

File: cppcache/test/TXIdTest.cpp (hunk 4 above)
Comment (pivotal-jbarrett) [reply cppcache/test/TXIdTest.cpp:29]: Use `std::unique_ptr` if this needs to even be a pointer. Probably just as testable as a stack allocated variable.

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

File: cppcache/test/TXIdTest.cpp (hunk 4 above)
Comment (jvarenina) [reply cppcache/test/TXIdTest.cpp:29]: Thanks. I have changed it to stack allocated variables.

**Review by pivotal-jbarrett — APPROVED** [review]

**Pull request merged**